### PR TITLE
Python3 compatibility

### PR DESCRIPTION
**Pull request by jadinm**
Mininet has python3 compatibility in its master branch for a few months.
This extends this compatibility to IPMininet.

### CODE REVIEW

File: ipmininet/iptopo.py [hunk 1]
```diff
@@ -116,7 +118,7 @@ def __new__(cls, value, *args, **kwargs):
 
     def __init__(self, o, topo):
         self.topo = topo
-        super(RouterDescription, self).__init__(o)
+        super(RouterDescription, self).__init__()
```
Comment (oliviertilmans) [inline ipmininet/iptopo.py]: Why did the init() parameter 'o' disappear here?

File: ipmininet/router/config/base.py [hunk 2]
```diff
@@ -2,6 +2,10 @@
 that is able to provide configurations for a set of routing daemons.
 It also defines the base class for a routing daemon, as well as a minimalistic
 configuration for a router."""
+from builtins import map
+from future.utils import iteritems
+from six import string_types
```
Comment (oliviertilmans) [inline ipmininet/router/config/base.py]: Not sure if pulling in six is really needed when all we want could be adding in the baselevel __init__.py=
```python
try:
  basestring
except NameError:
  basestring = str
```

File: ipmininet/link.py [hunk 3]
```diff
@@ -261,6 +264,29 @@ def __init__(self, node1, node2, intf=IPIntf, *args, **kwargs):
 TCIntf.__bases__ = (IPIntf,)
 
 
+class OrderedAddress(object):
```
Comment (oliviertilmans) [inline ipmininet/link.py]: You may want to look into [functools.total_ordering](https://docs.python.org/3/library/functools.html#functools.total_ordering) to dreduce the boilerplate there.

TLDR:
```python
@functools.total_orderding
class OrderedAddress(object):
    def __init__(self, addr):
        self.addr = addr

    def __lt__(self, other):
        return address_comparator(self.addr, other.addr) < 0

    def __eq__(self, other):
        return address_comparator(self.addr, other.addr) == 0
```

**Comment by oliviertilmans** [comment]
> Mininet has python3 compatibility in its master branch for a few months.
> This extends this compatibility to IPMininet.

It was always disappointing to see mininet lagging behind in py2 world, glad you could do this.

Did you try to run `2to3` on the codebase to check you did not miss 'obvious' uncompatibilities?

**Comment by jadinm** [comment]
> The changes are surprisingly few and look harmless.

We could reduce them even further by removing all the "from builtins import map, filter".
It would be slightly ineffective for py2 because list(map()) or list(fitler()) copy the list
but I do not think we will ever exceed the 100 items in these occurrences.
What do you think?

> Did you try to run `2to3` on the codebase to check you did not miss 'obvious' uncompatibilities?

I just checked and I forgot the metaclasses ^^
So, I added a new commit to the Pull Request.
The rest was already made.

File: ipmininet/iptopo.py (hunk 1 above)
Comment (jadinm) [reply ipmininet/iptopo.py]: Because Python3 raises the following exception if the argument is present:
```
TypeError: object.__init__() takes no parameters
```

Comment (oliviertilmans) [reply ipmininet/iptopo.py]: What are we missing as a functionnality due to that then?

Comment (jadinm) [reply ipmininet/iptopo.py]: None, I think. Python2.7 was already complaining about that in a warning. Python3 enforces it.
This creates a perfectly fine str object.

**Comment by oliviertilmans** [comment]
Then what was it doing there in the first place?
Should we store o as self.o and then override __str__ and __repr__ to return self.o (i.e. not subclass str anymore?)
________________________________

**Comment by jadinm** [comment]
> Then what was it doing there in the first place?

Passing ``o`` to the ``super().__init__()`` was pointless from the very beginning.
This method never took any argument.
But I had not noticed the warning issued by Python2 in the ocean of logs of Mininet.
I noticed it while coding this PR.

> Should we store o as self.o and then override __str__ and __repr__ to return self.o (i.e. not subclass str anymore?)

We don't need to do that.
Removing the parameter does not prevent the RouterDescription object of being used as a str object.
The following code
```
r1 = self.addRouter('r1')
print("%s - %s - %s" % (r1.count("r"), r1.count("a"), r1.upper()))
```
will print
```
1 - 0 - R1
```

Moreover the other tests are working fine as well.

**Comment by jadinm** [comment]
I suggest waiting for PR #18 before merging.
This way I would also check python3 compatibility for the new daemon as well.

**Comment by jadinm** [comment]
I extended the PR for OpenR and the installation scripts.

File: ipmininet/router/config/base.py [hunk 4]
```diff
@@ -35,7 +39,7 @@ def __init__(self, node, daemons=(), sysctl=None,
                        interfaces."""
         self._node = node  # The node for which we will build the configuration
         self._daemons = {}  # Active daemons
-        map(self.register_daemon, daemons)
+        list(map(self.register_daemon, daemons))
```
Comment (oliviertilmans) [inline ipmininet/router/config/base.py]: The map call there was a (pointless) way to make this a one-liner instead of a simple for loop.
You definitely do not need the list expansion (but still need to run the loop).

I'd rewrite all such case by their for loop equivalent ...

```python
for d in daemons:
    self.register_daemon(d)
```

File: ipmininet/router/config/base.py [hunk 5]
```diff
@@ -51,23 +55,23 @@ def build(self):
         self._cfg.password = self._node.password
         self._cfg.name = self._node.name
         # Check that all daemons have their dependencies satisfied
-        map(self.register_daemon,
-            (c for cls in self._daemons.values()
-             for c in cls.DEPENDS if c.NAME not in self._daemons))
+        list(map(self.register_daemon,
+                 (c for cls in list(self._daemons.values())
+                  for c in cls.DEPENDS if c.NAME not in self._daemons)))
```
Comment (oliviertilmans) [inline ipmininet/router/config/base.py]: Same than above (unneeded map in the first place)

File: ipmininet/router/config/bgp.py [hunk 6]
```diff
@@ -54,7 +56,7 @@ def bgp_fullmesh(topo, routers):
     :param routers: The set of routers peering within each other"""
     def _set_peering(x):
         bgp_peering(topo, x[0], x[1])
-    map(_set_peering, itertools.combinations(routers, 2))
+    list(map(_set_peering, itertools.combinations(routers, 2)))
```
Comment (oliviertilmans) [inline ipmininet/router/config/bgp.py]: unneeded map

File: ipmininet/router/config/bgp.py [hunk 7]
```diff
@@ -118,7 +120,7 @@ class AddressFamily(object):
     def __init__(self, af_name, redistribute=(), networks=(),
                  *args, **kwargs):
         self.name = af_name
-        self.networks = map(ip_network, map(unicode, networks))
+        self.networks = list(map(ip_network, map(str, networks)))
```
Comment (oliviertilmans) [inline ipmininet/router/config/bgp.py]: I'd like to avoid the double list expansion that happens in py2 were possible.
List comprehensions would fit the bill here:

```python
self.networks = [ip_network(str(n)) for n in networks]
```

File: ipmininet/link.py [hunk 8]
```diff
@@ -150,7 +154,7 @@ def _set_ip(self, ip, prefixLen=None):
                     addr = ip_interface(u'%s/%s' % (addr, prefixLen))
                 else:
                     # no prefixLen defaults to full /128 or /32
-                    addr = ip_interface(unicode(addr))
+                    addr = ip_interface(str(addr))
```
Comment (oliviertilmans) [inline ipmininet/link.py]: Did you test this in py2?

IIRC the [python2 ipaddress module backport](https://pypi.org/project/ipaddress/) will straight out [refuse non-unicode input strings](https://github.com/phihag/ipaddress/blob/master/README.md).

File: ipmininet/router/config/iptables.py [hunk 9]
```diff
@@ -38,7 +40,7 @@ def set_defaults(self, defaults):
     def build(self):
         cfg = super(IPTables, self).build()
         table_name = attrgetter('table')
-        cfg.rules = {k: map(str, v)
+        cfg.rules = {k: list(map(str, v))
```
Comment (oliviertilmans) [inline ipmininet/router/config/iptables.py]: Similar than below, rewrite this as a list comprehension instead

File: ipmininet/router/config/ospf.py [hunk 10]
```diff
@@ -30,7 +32,7 @@ def area(self, x):
     def apply(self, topo):
         # Add all links for the routers
         for r in self.nodes:
-            self.add_link(*map(lambda x: (r, x), topo.g[r].keys()))
+            self.add_link(*list(map(lambda x: (r, x), topo.g[r].keys())))
```
Comment (oliviertilmans) [inline ipmininet/router/config/ospf.py]:
```python
self.add_link(*((r, x) for x in top.g[r].keys()))
```

File: ipmininet/link.py (hunk 8 above)
Comment (jadinm) [reply ipmininet/link.py]: All the code was tested on Python 2.7, 3.5, 3.6 and 3.7.
This line is not covered by tests, but you have a similar modification in ipmininet/utils.py, line 77 which is covered by the tests.

The "str" is not the python2 "str". It is imported from "builtins" (which is installed with the future package) at the top of the file.
This function consistently returns a unicode string no matter the python version.

If you still need to be conviced, this code works in python2.7 and the latest version of ipaddress.
```python
import ipaddress
from builtins import str
ipaddress.ip_address(str(b"1.1.1.1"))
```
while the following doesn't
```
import ipaddress
ipaddress.ip_address(str(b"1.1.1.1"))
˘˘˘

Comment (oliviertilmans) [reply ipmininet/link.py]: I did not notice the str override through import.
What a nasty/clever monkey patch.
